{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import AlignIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Sequences: 14\n",
      "Alignment Length: 77\n",
      "GFGTYCPTTCGVADYLQRYKPDMDKKLDDMEQDLEEIANLTRGAQDKVVYLK---DSEAQAQKQSPDTYIKKSSNML Q7ZVG7_BRARE/37-110\n",
      "RFGSYCPTTCGIADFLSTYQATVDKDLQTLEDILSQAENKTMEAKELVKAIQVSYLPEDPARPNRVELATKDSKKMM Q6X871_SCAAQ/1-77\n",
      "RFGSYCPTTCGIADFLSTYQTGVXNDLRTLEDLLSGIENKTSEAKELIKSIQVSYNPNEPPKPNTIVSATKDSKKMM O02676_CROCR/1-77\n",
      "RFGSYCPTTCGIADFLSTYQGSIDKDLQTLEDILNQVENKTXEASELIKSIQVSYNPDEPPRPNMIEGATQKSKKML Q6X869_TENEC/1-77\n",
      "RFGSYCPTTCGIADFLSTYQTKVDKDLQSLEDILHQVENKTSEVKQLIKAIQLTYNPDESSKPNMIDAATLKSRKML FIBG_HUMAN/40-116\n",
      "RFGSYCPTTCGIADFLSTYQTXVDKDLQVLEDILNQAENKTSEAKELIKAIQVRYKPDEPTKPGGIDSATRESKKML O02689_TAPIN/1-77\n",
      "RFGSYCPTMCGIAGFLSTYQNTVEKDLQNLEGILHQVENKTSEARELIKAIQISYNPEDLSKPDRIQSATKESKKML O02688_PIG/1-77\n",
      "RFGSYCPTTCGVADFLSNYQTSVDKDLQNLEGILYQVENKTSEARELVKAIQISYNPDEPSKPNNIESATKNSKRMM O02672_9CETA/1-77\n",
      "RFGSYCPTTCGIADFLSNYQTSVDKDLQDFEDILHRAENQTSEAEQLIQAIRTSYNPDEPPKTGRIDAATRESKKMM O02682_EQUPR/1-77\n",
      "RFGSYCPTTCGIADFLSTYQTKVDEDLQNLEDILYRVENRTSEAKELIKAIQVDYNPGEPPKQSVTEGATQNAKKMV Q6X870_CYNVO/1-77\n",
      "RFGSYCPTTCGISDFLNSYQTDVDTDLQTLENILQRAENRTTEAKELIKAIQVYYNPDQPPKPGMIEGATQKSKKMV FIBG_RAT/40-116\n",
      "RFGSYCPTTCGIADFLNKYQTTIDQDLRHMEETLRDIDNKTAESTLLIQKIQIGQTPDPRPQ-NVIGDVTQKSRKMI Q6X866_DROAU/1-76\n",
      "RFGSYCPTTCGIADFFNKYRLTTDGELLEIEGLLQQATNSTGSIEYLIQHIKTIYPSEKQTLPQSIEQLTQKSKKII O93568_CHICK/40-116\n",
      "RFGEYCPTTCGISDFLNRYQENVDTDLQYLENLLTQISNSTSGTTIIVEHLIDSGKKPATSPQTAIDPMTQKSKTCW FIBG_XENLA/38-114\n"
     ]
    }
   ],
   "source": [
    "align = AlignIO.read(\"PF09395_seed.sth\", \"stockholm\") #filename, format\n",
    "print(\"# Sequences: {}\".format(len(align)))\n",
    "print(\"Alignment Length: {}\".format(align.get_alignment_length()))\n",
    "\n",
    "for record in align:\n",
    "    print(record.seq + \" \" + record.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      A     C     D     E     F     G   H     I     K     L    M     N     P     Q    R     S     T    V   W   X     Y\n",
      "A 197.0   0.0   3.5  12.5   0.0   4.5 8.0  15.0   5.0   6.5  7.0   1.0   8.5   2.0  2.0  45.0  13.5 23.5 0.0 0.5   9.0\n",
      "C   0.0 182.0   0.0   0.0   0.0   0.0 0.0   0.5   0.0   0.0  6.0   0.0   0.0   0.0  0.0   0.0   0.0  0.0 0.0 0.0   0.0\n",
      "D   3.5   0.0 266.0  65.0   0.0  28.0 1.0   7.0  17.0   4.0  0.0  23.5   5.5  18.5  3.0  10.5  12.0  6.5 0.0 7.0   2.0\n",
      "E  12.5   0.0  65.0 265.0   0.0  19.0 2.0   4.0  29.0   4.5  0.0   9.5   6.0  23.0  4.0  19.0  15.0  4.0 0.0 0.5   5.5\n",
      "F   0.0   0.0   0.0   0.0 169.0   0.0 0.0   0.5   0.0  11.5  1.0   0.0   0.0   0.0  0.0   0.0   0.0  0.0 0.0 0.0   6.5\n",
      "G   4.5   0.0  28.0  19.0   0.0 192.0 0.0   0.5   5.0   2.0  2.0  14.5   2.5  11.5 10.5  18.5  11.5  1.5 0.0 1.0   6.0\n",
      "H   8.0   0.0   1.0   2.0   0.0   0.0 4.0   0.0   1.0   0.0  0.0   4.5   0.0   3.0  1.5   5.5   3.5  0.5 0.0 0.0   4.5\n",
      "I  15.0   0.5   7.0   4.0   0.5   0.5 0.0 279.0   1.5  43.5 12.0   0.0   5.5   6.0  1.0   4.0  21.5 69.0 1.0 0.0   1.0\n",
      "K   5.0   0.0  17.0  29.0   0.0   5.0 1.0   1.5 280.0  17.5  0.0  27.5  15.5  65.0 47.0  22.5  32.0  4.5 0.0 1.0   0.5\n",
      "L   6.5   0.0   4.0   4.5  11.5   2.0 0.0  43.5  17.5 385.0 20.5   4.5   6.0  12.0  4.5   3.5   8.0  8.5 2.5 0.0   1.0\n",
      "M   7.0   6.0   0.0   0.0   1.0   2.0 0.0  12.0   0.0  20.5 76.0   1.5   0.0   0.0  5.0   7.0   9.0 12.5 2.0 0.5   0.5\n",
      "N   1.0   0.0  23.5   9.5   0.0  14.5 4.5   0.0  27.5   4.5  1.5 147.0   5.0  11.5  5.0  28.0  29.0  2.5 0.0 0.5   3.5\n",
      "P   8.5   0.0   5.5   6.0   0.0   2.5 0.0   5.5  15.5   6.0  0.0   5.0 225.0  19.0  5.0  29.0  20.0  0.5 0.0 0.0   0.0\n",
      "Q   2.0   0.0  18.5  23.0   0.0  11.5 3.0   6.0  65.0  12.0  0.0  11.5  19.0 208.0 45.0  12.0   6.5  2.0 0.0 0.0   8.5\n",
      "R   2.0   0.0   3.0   4.0   0.0  10.5 1.5   1.0  47.0   4.5  5.0   5.0   5.0  45.0 91.0  17.0  13.0  3.0 0.0 0.5   1.5\n",
      "S  45.0   0.0  10.5  19.0   0.0  18.5 5.5   4.0  22.5   3.5  7.0  28.0  29.0  12.0 17.0 250.0  36.0  3.0 0.0 5.5   7.0\n",
      "T  13.5   0.0  12.0  15.0   0.0  11.5 3.5  21.5  32.0   8.0  9.0  29.0  20.0   6.5 13.0  36.0 403.0 15.0 0.0 2.5   3.5\n",
      "V  23.5   0.0   6.5   4.0   0.0   1.5 0.5  69.0   4.5   8.5 12.5   2.5   0.5   2.0  3.0   3.0  15.0 79.0 1.0 0.0   1.0\n",
      "W   0.0   0.0   0.0   0.0   0.0   0.0 0.0   1.0   0.0   2.5  2.0   0.0   0.0   0.0  0.0   0.0   0.0  1.0 0.0 0.0   0.0\n",
      "X   0.5   0.0   7.0   0.5   0.0   1.0 0.0   0.0   1.0   0.0  0.5   0.5   0.0   0.0  0.5   5.5   2.5  0.0 0.0 0.0   0.0\n",
      "Y   9.0   0.0   2.0   5.5   6.5   6.0 4.5   1.0   0.5   1.0  0.5   3.5   0.0   8.5  1.5   7.0   3.5  1.0 0.0 0.0 238.0\n",
      "\n",
      "     A    C    D    E    F    G   H    I    K    L   M    N    P    Q   R    S    T   V   W   X    Y\n",
      "A 14.1  0.0  0.2  0.9  0.0  0.3 0.6  1.1  0.4  0.5 0.5  0.1  0.6  0.1 0.1  3.2  1.0 1.7 0.0 0.0  0.6\n",
      "C  0.0 13.0  0.0  0.0  0.0  0.0 0.0  0.0  0.0  0.0 0.4  0.0  0.0  0.0 0.0  0.0  0.0 0.0 0.0 0.0  0.0\n",
      "D  0.2  0.0 19.0  4.6  0.0  2.0 0.1  0.5  1.2  0.3 0.0  1.7  0.4  1.3 0.2  0.8  0.9 0.5 0.0 0.5  0.1\n",
      "E  0.9  0.0  4.6 18.9  0.0  1.4 0.1  0.3  2.1  0.3 0.0  0.7  0.4  1.6 0.3  1.4  1.1 0.3 0.0 0.0  0.4\n",
      "F  0.0  0.0  0.0  0.0 12.1  0.0 0.0  0.0  0.0  0.8 0.1  0.0  0.0  0.0 0.0  0.0  0.0 0.0 0.0 0.0  0.5\n",
      "G  0.3  0.0  2.0  1.4  0.0 13.7 0.0  0.0  0.4  0.1 0.1  1.0  0.2  0.8 0.8  1.3  0.8 0.1 0.0 0.1  0.4\n",
      "H  0.6  0.0  0.1  0.1  0.0  0.0 0.3  0.0  0.1  0.0 0.0  0.3  0.0  0.2 0.1  0.4  0.2 0.0 0.0 0.0  0.3\n",
      "I  1.1  0.0  0.5  0.3  0.0  0.0 0.0 19.9  0.1  3.1 0.9  0.0  0.4  0.4 0.1  0.3  1.5 4.9 0.1 0.0  0.1\n",
      "K  0.4  0.0  1.2  2.1  0.0  0.4 0.1  0.1 20.0  1.2 0.0  2.0  1.1  4.6 3.4  1.6  2.3 0.3 0.0 0.1  0.0\n",
      "L  0.5  0.0  0.3  0.3  0.8  0.1 0.0  3.1  1.2 27.5 1.5  0.3  0.4  0.9 0.3  0.2  0.6 0.6 0.2 0.0  0.1\n",
      "M  0.5  0.4  0.0  0.0  0.1  0.1 0.0  0.9  0.0  1.5 5.4  0.1  0.0  0.0 0.4  0.5  0.6 0.9 0.1 0.0  0.0\n",
      "N  0.1  0.0  1.7  0.7  0.0  1.0 0.3  0.0  2.0  0.3 0.1 10.5  0.4  0.8 0.4  2.0  2.1 0.2 0.0 0.0  0.2\n",
      "P  0.6  0.0  0.4  0.4  0.0  0.2 0.0  0.4  1.1  0.4 0.0  0.4 16.1  1.4 0.4  2.1  1.4 0.0 0.0 0.0  0.0\n",
      "Q  0.1  0.0  1.3  1.6  0.0  0.8 0.2  0.4  4.6  0.9 0.0  0.8  1.4 14.9 3.2  0.9  0.5 0.1 0.0 0.0  0.6\n",
      "R  0.1  0.0  0.2  0.3  0.0  0.8 0.1  0.1  3.4  0.3 0.4  0.4  0.4  3.2 6.5  1.2  0.9 0.2 0.0 0.0  0.1\n",
      "S  3.2  0.0  0.8  1.4  0.0  1.3 0.4  0.3  1.6  0.2 0.5  2.0  2.1  0.9 1.2 17.9  2.6 0.2 0.0 0.4  0.5\n",
      "T  1.0  0.0  0.9  1.1  0.0  0.8 0.2  1.5  2.3  0.6 0.6  2.1  1.4  0.5 0.9  2.6 28.8 1.1 0.0 0.2  0.2\n",
      "V  1.7  0.0  0.5  0.3  0.0  0.1 0.0  4.9  0.3  0.6 0.9  0.2  0.0  0.1 0.2  0.2  1.1 5.6 0.1 0.0  0.1\n",
      "W  0.0  0.0  0.0  0.0  0.0  0.0 0.0  0.1  0.0  0.2 0.1  0.0  0.0  0.0 0.0  0.0  0.0 0.1 0.0 0.0  0.0\n",
      "X  0.0  0.0  0.5  0.0  0.0  0.1 0.0  0.0  0.1  0.0 0.0  0.0  0.0  0.0 0.0  0.4  0.2 0.0 0.0 0.0  0.0\n",
      "Y  0.6  0.0  0.1  0.4  0.5  0.4 0.3  0.1  0.0  0.1 0.0  0.2  0.0  0.6 0.1  0.5  0.2 0.1 0.0 0.0 17.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#how often do different substitutions occur?\n",
    "print(align.substitutions)\n",
    "#amino acid stays in A, 197 times out of all of these pairwise comparisons\n",
    "#T changed to M 9 times\n",
    "print(align.substitutions/len(align))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">Q7ZVG7_BRARE/37-110\n",
      "GFGTYCPTTCGVADYLQRYKPDMDKKLDDMEQDLEEIANLTRGAQDKVVYLK---DSEAQ\n",
      "AQKQSPDTYIKKSSNML\n",
      ">Q6X871_SCAAQ/1-77\n",
      "RFGSYCPTTCGIADFLSTYQATVDKDLQTLEDILSQAENKTMEAKELVKAIQVSYLPEDP\n",
      "ARPNRVELATKDSKKMM\n",
      ">O02676_CROCR/1-77\n",
      "RFGSYCPTTCGIADFLSTYQTGVXNDLRTLEDLLSGIENKTSEAKELIKSIQVSYNPNEP\n",
      "PKPNTIVSATKDSKKMM\n",
      ">Q6X869_TENEC/1-77\n",
      "RFGSYCPTTCGIADFLSTYQGSIDKDLQTLEDILNQVENKTXEASELIKSIQVSYNPDEP\n",
      "PRPNMIEGATQKSKKML\n",
      ">FIBG_HUMAN/40-116\n",
      "RFGSYCPTTCGIADFLSTYQTKVDKDLQSLEDILHQVENKTSEVKQLIKAIQLTYNPDES\n",
      "SKPNMIDAATLKSRKML\n",
      ">O02689_TAPIN/1-77\n",
      "RFGSYCPTTCGIADFLSTYQTXVDKDLQVLEDILNQAENKTSEAKELIKAIQVRYKPDEP\n",
      "TKPGGIDSATRESKKML\n",
      ">O02688_PIG/1-77\n",
      "RFGSYCPTMCGIAGFLSTYQNTVEKDLQNLEGILHQVENKTSEARELIKAIQISYNPEDL\n",
      "SKPDRIQSATKESKKML\n",
      ">O02672_9CETA/1-77\n",
      "RFGSYCPTTCGVADFLSNYQTSVDKDLQNLEGILYQVENKTSEARELVKAIQISYNPDEP\n",
      "SKPNNIESATKNSKRMM\n",
      ">O02682_EQUPR/1-77\n",
      "RFGSYCPTTCGIADFLSNYQTSVDKDLQDFEDILHRAENQTSEAEQLIQAIRTSYNPDEP\n",
      "PKTGRIDAATRESKKMM\n",
      ">Q6X870_CYNVO/1-77\n",
      "RFGSYCPTTCGIADFLSTYQTKVDEDLQNLEDILYRVENRTSEAKELIKAIQVDYNPGEP\n",
      "PKQSVTEGATQNAKKMV\n",
      ">FIBG_RAT/40-116\n",
      "RFGSYCPTTCGISDFLNSYQTDVDTDLQTLENILQRAENRTTEAKELIKAIQVYYNPDQP\n",
      "PKPGMIEGATQKSKKMV\n",
      ">Q6X866_DROAU/1-76\n",
      "RFGSYCPTTCGIADFLNKYQTTIDQDLRHMEETLRDIDNKTAESTLLIQKIQIGQTPDPR\n",
      "PQ-NVIGDVTQKSRKMI\n",
      ">O93568_CHICK/40-116\n",
      "RFGSYCPTTCGIADFFNKYRLTTDGELLEIEGLLQQATNSTGSIEYLIQHIKTIYPSEKQ\n",
      "TLPQSIEQLTQKSKKII\n",
      ">FIBG_XENLA/38-114\n",
      "RFGEYCPTTCGISDFLNRYQENVDTDLQYLENLLTQISNSTSGTTIIVEHLIDSGKKPAT\n",
      "SPQTAIDPMTQKSKTCW\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#convert file format\n",
    "print(align.format(\"fasta\"))\n",
    "AlignIO.convert(\"PF09395_seed.sth\", \"stockholm\", \"PF09395_seed.phy\", \"phylip\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Program: blastn (2.2.27+)\n",
      "  Query: 42291 (61)\n",
      "         mystery_seq\n",
      " Target: refseq_rna\n",
      "   Hits: ----  -----  ----------------------------------------------------------\n",
      "            #  # HSP  ID + description\n",
      "         ----  -----  ----------------------------------------------------------\n",
      "            0      1  gi|262205317|ref|NR_030195.1|  Homo sapiens microRNA 52...\n",
      "            1      1  gi|301171311|ref|NR_035856.1|  Pan troglodytes microRNA...\n",
      "            2      1  gi|270133242|ref|NR_032573.1|  Macaca mulatta microRNA ...\n",
      "            3      2  gi|301171322|ref|NR_035857.1|  Pan troglodytes microRNA...\n",
      "            4      1  gi|301171267|ref|NR_035851.1|  Pan troglodytes microRNA...\n",
      "            5      2  gi|262205330|ref|NR_030198.1|  Homo sapiens microRNA 52...\n",
      "            6      1  gi|262205302|ref|NR_030191.1|  Homo sapiens microRNA 51...\n",
      "            7      1  gi|301171259|ref|NR_035850.1|  Pan troglodytes microRNA...\n",
      "            8      1  gi|262205451|ref|NR_030222.1|  Homo sapiens microRNA 51...\n",
      "            9      2  gi|301171447|ref|NR_035871.1|  Pan troglodytes microRNA...\n",
      "           10      1  gi|301171276|ref|NR_035852.1|  Pan troglodytes microRNA...\n",
      "           11      1  gi|262205290|ref|NR_030188.1|  Homo sapiens microRNA 51...\n",
      "           12      1  gi|301171354|ref|NR_035860.1|  Pan troglodytes microRNA...\n",
      "           13      1  gi|262205281|ref|NR_030186.1|  Homo sapiens microRNA 52...\n",
      "           14      2  gi|262205298|ref|NR_030190.1|  Homo sapiens microRNA 52...\n",
      "           15      1  gi|301171394|ref|NR_035865.1|  Pan troglodytes microRNA...\n",
      "           16      1  gi|262205429|ref|NR_030218.1|  Homo sapiens microRNA 51...\n",
      "           17      1  gi|262205423|ref|NR_030217.1|  Homo sapiens microRNA 52...\n",
      "           18      1  gi|301171401|ref|NR_035866.1|  Pan troglodytes microRNA...\n",
      "           19      1  gi|270133247|ref|NR_032574.1|  Macaca mulatta microRNA ...\n",
      "           20      1  gi|262205309|ref|NR_030193.1|  Homo sapiens microRNA 52...\n",
      "           21      2  gi|270132717|ref|NR_032716.1|  Macaca mulatta microRNA ...\n",
      "           22      2  gi|301171437|ref|NR_035870.1|  Pan troglodytes microRNA...\n",
      "           23      2  gi|270133306|ref|NR_032587.1|  Macaca mulatta microRNA ...\n",
      "           24      2  gi|301171428|ref|NR_035869.1|  Pan troglodytes microRNA...\n",
      "           25      1  gi|301171211|ref|NR_035845.1|  Pan troglodytes microRNA...\n",
      "           26      2  gi|301171153|ref|NR_035838.1|  Pan troglodytes microRNA...\n",
      "           27      2  gi|301171146|ref|NR_035837.1|  Pan troglodytes microRNA...\n",
      "           28      2  gi|270133254|ref|NR_032575.1|  Macaca mulatta microRNA ...\n",
      "           29      2  gi|262205445|ref|NR_030221.1|  Homo sapiens microRNA 51...\n",
      "           ~~~\n",
      "           97      1  gi|356517317|ref|XM_003527287.1|  PREDICTED: Glycine ma...\n",
      "           98      1  gi|297814701|ref|XM_002875188.1|  Arabidopsis lyrata su...\n",
      "           99      1  gi|397513516|ref|XM_003827011.1|  PREDICTED: Pan panisc...\n"
     ]
    }
   ],
   "source": [
    "#BLAST sequence search tool finds regions of similarity between biological sequences\n",
    "#search all of the databases from NCBI\n",
    "#rna, dna, protein sequences\n",
    "#returns a file that tells you how the search results matched up with your query\n",
    "#what database records hit, the name of that record, and the sequence and how your sequence aligns with it\n",
    "#my_blast.xml is an example search result file\n",
    "\n",
    "#Bio.SearchIO provides methods for reading these search results\n",
    "#list of high scoring segment pairs\n",
    "\n",
    "from Bio import SearchIO\n",
    "result = SearchIO.read(\"my_blast.xml\", \"blast-xml\")\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query: 42291\n",
      "       mystery_seq\n",
      "  Hit: gi|262205317|ref|NR_030195.1| (61)\n",
      "       Homo sapiens microRNA 520b (MIR520B), microRNA\n",
      " HSPs: ----  --------  ---------  ------  ---------------  ---------------------\n",
      "          #   E-value  Bit score    Span      Query range              Hit range\n",
      "       ----  --------  ---------  ------  ---------------  ---------------------\n",
      "          0   4.9e-23     111.29      61           [0:61]                 [0:61]\n",
      "Query: 42291\n",
      "       mystery_seq\n",
      "  Hit: gi|301171311|ref|NR_035856.1| (60)\n",
      "       Pan troglodytes microRNA mir-520b (MIR520B), microRNA\n",
      " HSPs: ----  --------  ---------  ------  ---------------  ---------------------\n",
      "          #   E-value  Bit score    Span      Query range              Hit range\n",
      "       ----  --------  ---------  ------  ---------------  ---------------------\n",
      "          0   1.7e-22     109.49      60           [1:61]                 [0:60]\n",
      "Query: 42291\n",
      "       mystery_seq\n",
      "  Hit: gi|270133242|ref|NR_032573.1| (85)\n",
      "       Macaca mulatta microRNA mir-519a (MIR519A), microRNA\n",
      " HSPs: ----  --------  ---------  ------  ---------------  ---------------------\n",
      "          #   E-value  Bit score    Span      Query range              Hit range\n",
      "       ----  --------  ---------  ------  ---------------  ---------------------\n",
      "          0   2.5e-20     102.28      61           [0:61]                [12:73]\n",
      "Query: 42291\n",
      "       mystery_seq\n",
      "  Hit: gi|301171322|ref|NR_035857.1| (86)\n",
      "       Pan troglodytes microRNA mir-520c (MIR520C), microRNA\n",
      " HSPs: ----  --------  ---------  ------  ---------------  ---------------------\n",
      "          #   E-value  Bit score    Span      Query range              Hit range\n",
      "       ----  --------  ---------  ------  ---------------  ---------------------\n",
      "          0   8.9e-20     100.47      60           [1:61]                [13:73]\n",
      "          1   3.3e-06      55.39      60           [0:60]                [13:73]\n",
      "Query: 42291\n",
      "       mystery_seq\n",
      "  Hit: gi|301171267|ref|NR_035851.1| (80)\n",
      "       Pan troglodytes microRNA mir-519b (MIR519B), microRNA\n",
      " HSPs: ----  --------  ---------  ------  ---------------  ---------------------\n",
      "          #   E-value  Bit score    Span      Query range              Hit range\n",
      "       ----  --------  ---------  ------  ---------------  ---------------------\n",
      "          0   8.9e-20     100.47      60           [0:60]                 [9:69]\n"
     ]
    }
   ],
   "source": [
    "for hit in result.hits[:5]:\n",
    "    print(hit)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">42291 mystery_seq\n",
      "CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTTTAGAGG\n",
      "G\n",
      ">gi|262205317|ref|NR_030195.1| Homo sapiens microRNA 520b (MIR520B), microRNA\n",
      "CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTTTAGAGG\n",
      "G\n",
      "\n",
      ">42291 mystery_seq\n",
      "CCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTTTAGAGGG\n",
      ">gi|301171311|ref|NR_035856.1| Pan troglodytes microRNA mir-520b (MIR520B), microRNA\n",
      "CCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTTTAGAGGG\n",
      "\n",
      ">42291 mystery_seq\n",
      "CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTTTAGAGG\n",
      "G\n",
      ">gi|270133242|ref|NR_032573.1| Macaca mulatta microRNA mir-519a (MIR519A), microRNA\n",
      "CCCTCTAGAGGGAAGCGCTTTCTGTGGTCTGAAAGAAAAGAAAGTGCTTCCTTTTAGAGG\n",
      "G\n",
      "\n",
      ">42291 mystery_seq\n",
      "CCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTTTAGAGGG\n",
      ">gi|301171322|ref|NR_035857.1| Pan troglodytes microRNA mir-520c (MIR520C), microRNA\n",
      "CCTCTAGAGGGAAGCACTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTTTAGAGGG\n",
      "\n",
      ">42291 mystery_seq\n",
      "CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTTTAGAGG\n",
      ">gi|301171322|ref|NR_035857.1| Pan troglodytes microRNA mir-520c (MIR520C), microRNA\n",
      "CCCTCTAAAAGGAAGCACTTTCTTTTCTTTCAGACAACAGAAAGTGCTTCCCTCTAGAGG\n",
      "\n",
      ">42291 mystery_seq\n",
      "CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTTTAGAGG\n",
      ">gi|301171267|ref|NR_035851.1| Pan troglodytes microRNA mir-519b (MIR519B), microRNA\n",
      "CCCTCTAGAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCATCCTTTTAGAGG\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for hit in result.hits[:5]:\n",
    "    for hsp in hit.hsps:\n",
    "        print(hsp.aln.format(\"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query: mystery_seq\n",
      "       <unknown description>\n",
      "  Hit: chr19 (59128983)\n",
      "       <unknown description>\n",
      " HSPs: ----  --------  ---------  ------  ---------------  ---------------------\n",
      "          #   E-value  Bit score    Span      Query range              Hit range\n",
      "       ----  --------  ---------  ------  ---------------  ---------------------\n",
      "          0         ?          ?       ?           [0:61]    [54204480:54204541]\n",
      "          1         ?          ?       ?           [0:61]    [54233104:54264463]\n",
      "          2         ?          ?       ?           [0:61]    [54254477:54260071]\n",
      "          3         ?          ?       ?           [1:61]    [54210720:54210780]\n",
      "          4         ?          ?       ?           [0:60]    [54198476:54198536]\n",
      "          5         ?          ?       ?           [0:61]    [54265610:54265671]\n",
      "          6         ?          ?       ?           [0:61]    [54238143:54240175]\n",
      "          7         ?          ?       ?           [0:60]    [54189735:54189795]\n",
      "          8         ?          ?       ?           [0:61]    [54185425:54185486]\n",
      "          9         ?          ?       ?           [0:60]    [54197657:54197717]\n",
      "         10         ?          ?       ?           [0:61]    [54255662:54255723]\n",
      "         11         ?          ?       ?           [0:61]    [54201651:54201712]\n",
      "         12         ?          ?       ?           [8:60]    [54206009:54206061]\n",
      "         13         ?          ?       ?          [10:61]    [54178987:54179038]\n",
      "         14         ?          ?       ?           [8:61]    [54212018:54212071]\n",
      "         15         ?          ?       ?           [8:51]    [54234278:54234321]\n",
      "         16         ?          ?       ?           [8:61]    [54238143:54238196]\n",
      "      Query: mystery_seq <unknown description>\n",
      "        Hit: chr19 <unknown description>\n",
      "Query range: [0:61] (1)\n",
      "  Hit range: [54204480:54204541] (1)\n",
      "Quick stats: evalue ?; bitscore ?\n",
      "  Fragments: 1 (? columns)\n",
      "      Query: mystery_seq <unknown description>\n",
      "        Hit: chr19 <unknown description>\n",
      "Query range: [0:61] (1)\n",
      "  Hit range: [54233104:54264463] (1)\n",
      "Quick stats: evalue ?; bitscore ?\n",
      "  Fragments: ---  --------------  ----------------------  ----------------------\n",
      "               #            Span             Query range               Hit range\n",
      "             ---  --------------  ----------------------  ----------------------\n",
      "               0               ?                  [0:18]     [54233104:54233122]\n",
      "               1               ?                 [18:61]     [54264420:54264463]\n",
      "      Query: mystery_seq <unknown description>\n",
      "        Hit: chr19 <unknown description>\n",
      "Query range: [0:61] (1)\n",
      "  Hit range: [54254477:54260071] (1)\n",
      "Quick stats: evalue ?; bitscore ?\n",
      "  Fragments: ---  --------------  ----------------------  ----------------------\n",
      "               #            Span             Query range               Hit range\n",
      "             ---  --------------  ----------------------  ----------------------\n",
      "               0               ?                  [0:18]     [54254477:54254495]\n",
      "               1               ?                 [18:61]     [54260028:54260071]\n",
      "      Query: mystery_seq <unknown description>\n",
      "        Hit: chr19 <unknown description>\n",
      "Query range: [1:61] (1)\n",
      "  Hit range: [54210720:54210780] (1)\n",
      "Quick stats: evalue ?; bitscore ?\n",
      "  Fragments: 1 (? columns)\n",
      "      Query: mystery_seq <unknown description>\n",
      "        Hit: chr19 <unknown description>\n",
      "Query range: [0:60] (1)\n",
      "  Hit range: [54198476:54198536] (1)\n",
      "Quick stats: evalue ?; bitscore ?\n",
      "  Fragments: 1 (? columns)\n",
      "      Query: mystery_seq <unknown description>\n",
      "        Hit: chr19 <unknown description>\n",
      "Query range: [0:61] (1)\n",
      "  Hit range: [54265610:54265671] (1)\n",
      "Quick stats: evalue ?; bitscore ?\n",
      "  Fragments: 1 (? columns)\n",
      "      Query: mystery_seq <unknown description>\n",
      "        Hit: chr19 <unknown description>\n",
      "Query range: [0:61] (1)\n",
      "  Hit range: [54238143:54240175] (1)\n",
      "Quick stats: evalue ?; bitscore ?\n",
      "  Fragments: ---  --------------  ----------------------  ----------------------\n",
      "               #            Span             Query range               Hit range\n",
      "             ---  --------------  ----------------------  ----------------------\n",
      "               0               ?                  [0:34]     [54238143:54238177]\n",
      "               1               ?                 [34:61]     [54240148:54240175]\n",
      "      Query: mystery_seq <unknown description>\n",
      "        Hit: chr19 <unknown description>\n",
      "Query range: [0:60] (1)\n",
      "  Hit range: [54189735:54189795] (1)\n",
      "Quick stats: evalue ?; bitscore ?\n",
      "  Fragments: 1 (? columns)\n",
      "      Query: mystery_seq <unknown description>\n",
      "        Hit: chr19 <unknown description>\n",
      "Query range: [0:61] (1)\n",
      "  Hit range: [54185425:54185486] (1)\n",
      "Quick stats: evalue ?; bitscore ?\n",
      "  Fragments: 1 (? columns)\n",
      "      Query: mystery_seq <unknown description>\n",
      "        Hit: chr19 <unknown description>\n",
      "Query range: [0:60] (1)\n",
      "  Hit range: [54197657:54197717] (1)\n",
      "Quick stats: evalue ?; bitscore ?\n",
      "  Fragments: 1 (? columns)\n",
      "      Query: mystery_seq <unknown description>\n",
      "        Hit: chr19 <unknown description>\n",
      "Query range: [0:61] (1)\n",
      "  Hit range: [54255662:54255723] (1)\n",
      "Quick stats: evalue ?; bitscore ?\n",
      "  Fragments: 1 (? columns)\n",
      "      Query: mystery_seq <unknown description>\n",
      "        Hit: chr19 <unknown description>\n",
      "Query range: [0:61] (1)\n",
      "  Hit range: [54201651:54201712] (1)\n",
      "Quick stats: evalue ?; bitscore ?\n",
      "  Fragments: 1 (? columns)\n",
      "      Query: mystery_seq <unknown description>\n",
      "        Hit: chr19 <unknown description>\n",
      "Query range: [8:60] (1)\n",
      "  Hit range: [54206009:54206061] (1)\n",
      "Quick stats: evalue ?; bitscore ?\n",
      "  Fragments: 1 (? columns)\n",
      "      Query: mystery_seq <unknown description>\n",
      "        Hit: chr19 <unknown description>\n",
      "Query range: [10:61] (1)\n",
      "  Hit range: [54178987:54179038] (1)\n",
      "Quick stats: evalue ?; bitscore ?\n",
      "  Fragments: 1 (? columns)\n",
      "      Query: mystery_seq <unknown description>\n",
      "        Hit: chr19 <unknown description>\n",
      "Query range: [8:61] (1)\n",
      "  Hit range: [54212018:54212071] (1)\n",
      "Quick stats: evalue ?; bitscore ?\n",
      "  Fragments: ---  --------------  ----------------------  ----------------------\n",
      "               #            Span             Query range               Hit range\n",
      "             ---  --------------  ----------------------  ----------------------\n",
      "               0               ?                  [8:50]     [54212018:54212060]\n",
      "               1               ?                 [50:54]     [54212061:54212065]\n",
      "               2               ?                 [55:61]     [54212065:54212071]\n",
      "      Query: mystery_seq <unknown description>\n",
      "        Hit: chr19 <unknown description>\n",
      "Query range: [8:51] (1)\n",
      "  Hit range: [54234278:54234321] (1)\n",
      "Quick stats: evalue ?; bitscore ?\n",
      "  Fragments: 1 (? columns)\n",
      "      Query: mystery_seq <unknown description>\n",
      "        Hit: chr19 <unknown description>\n",
      "Query range: [8:61] (-1)\n",
      "  Hit range: [54238143:54238196] (1)\n",
      "Quick stats: evalue ?; bitscore ?\n",
      "  Fragments: 1 (? columns)\n"
     ]
    }
   ],
   "source": [
    "blatresult = SearchIO.read(\"my_blat.psl\", \"blat-psl\")\n",
    "for hit in blatresult.hits[:5]:\n",
    "    print(hit)\n",
    "    for hsp in hit.hsps:\n",
    "        print(hsp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1, 100, 137, 137)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SearchIO.convert(\"my_blast.xml\", \"blast-xml\", \"my_blast.txt\", \"blast-tab\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
